### PR TITLE
[feat]: [CDS-38353]: Email Step Skeleton Code

### PR DESCRIPTION
**Pull request by vivekkarnatiharness**
## Describe your changes
Have written all the boilerplate code for the email step

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/34845)

### CODE REVIEW

File: 860-orchestration-steps/src/main/java/io/harness/plancreator/steps/email/EmailStepPlanCreator.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+import io.harness.plancreator.steps.internal.PMSStepPlanCreatorV2;
+import io.harness.pms.sdk.core.plan.creation.beans.PlanCreationContext;
+import io.harness.pms.sdk.core.plan.creation.beans.PlanCreationResponse;
+import io.harness.steps.StepSpecTypeConstants;
+
+import com.google.common.collect.Sets;
+import java.util.Set;
+
+public class EmailStepPlanCreator extends PMSStepPlanCreatorV2<EmailStepNode> {
```
Comment (puthrayaharness) [inline 860-orchestration-steps/src/main/java/io/harness/plancreator/steps/email/EmailStepPlanCreator.java:11]: Add all relevant annotations. e.g owned by

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/email/EmailStepResponse.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package io.harness.delegate.task.email;
+
+import io.harness.tasks.ResponseData;
+
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+public class EmailStepResponse implements ResponseData {}
```
Comment (puthrayaharness) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/email/EmailStepResponse.java:10]: Add owned by

File: 860-orchestration-steps/src/main/java/io/harness/plancreator/steps/email/EmailStepNode.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import lombok.NoArgsConstructor;
+import org.springframework.data.annotation.TypeAlias;
+
+@Data
+@NoArgsConstructor
+@EqualsAndHashCode(callSuper = true)
+@JsonTypeName(StepSpecTypeConstants.EMAIL)
+@TypeAlias("EmailStepNode")
+@OwnedBy(CDC)
+@RecasterAlias("io.harness.steps.email.EmailStepNode")
```
Comment (puthrayaharness) [inline 860-orchestration-steps/src/main/java/io/harness/plancreator/steps/email/EmailStepNode.java:30]: This is not correct. I think it should be `io.harness.plancreator.steps.email.EmailStepNode`

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

**Comment by vivekkarnatiharness** [comment]
trigger messagecheck

**Comment by vivekkarnatiharness** [comment]
trigger messagecheck

File: 860-orchestration-steps/src/main/java/io/harness/plancreator/steps/email/EmailStepPlanCreator.java (hunk 1 above)
Comment (vivekkarnatiharness) [reply 860-orchestration-steps/src/main/java/io/harness/plancreator/steps/email/EmailStepPlanCreator.java:11]: In shellscriptstepplancreator,httpstepplancreator there is no annoations added.
So i have added only Ownedby(CDC) right now...

File: 860-orchestration-steps/src/main/java/io/harness/serializer/kryo/OrchestrationStepsKryoRegistrar.java [hunk 4]
```diff
@@ -71,5 +71,8 @@ public void register(Kryo kryo) {
     kryo.register(HttpStepInfo.class, 8048);
     kryo.register(HttpOutcome.class, 12501);
     kryo.register(ShellScriptStepInfo.class, 8055);
+
+    //    kryo.register(EmailStepInfo.class, 13001);
```
[the remark this answers is not in the document]
Comment (vivekkarnatiharness) [reply 860-orchestration-steps/src/main/java/io/harness/serializer/kryo/OrchestrationStepsKryoRegistrar.java]: Removed

File: 860-orchestration-steps/src/main/java/io/harness/steps/Email/EmailOutcome.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import io.harness.annotations.dev.OwnedBy;
+import io.harness.pms.sdk.core.data.Outcome;
+
+import lombok.Builder;
+import lombok.Value;
+
+@Value
+@Builder
+@OwnedBy(CDC)
+@RecasterAlias("io.harness.steps.Email.EmailOutcome")
```
[the remark this answers is not in the document]
Comment (vivekkarnatiharness) [reply 860-orchestration-steps/src/main/java/io/harness/steps/Email/EmailOutcome.java]: Done

File: 860-orchestration-steps/src/main/java/io/harness/steps/Email/EmailStep.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import io.harness.pms.contracts.steps.StepType;
+import io.harness.pms.sdk.core.steps.io.StepInputPackage;
+import io.harness.pms.sdk.core.steps.io.StepResponse;
+import io.harness.serializer.KryoSerializer;
+import io.harness.steps.StepSpecTypeConstants;
+import io.harness.supplier.ThrowingSupplier;
+
+import com.google.inject.Inject;
+
+public class EmailStep extends TaskExecutableWithRollback<EmailStepResponse> {
```
[the remark this answers is not in the document]
Comment (vivekkarnatiharness) [reply 860-orchestration-steps/src/main/java/io/harness/steps/Email/EmailStep.java]: Done

File: 860-orchestration-steps/src/main/java/io/harness/steps/Email/EmailStepParameters.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import org.springframework.data.annotation.TypeAlias;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@TypeAlias("EmailStepParameters")
+@Builder
+public class EmailStepParameters implements SpecParameters {
```
[the remark this answers is not in the document]
Comment (vivekkarnatiharness) [reply 860-orchestration-steps/src/main/java/io/harness/steps/Email/EmailStepParameters.java]: Done

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/email/EmailStepResponse.java (hunk 2 above)
Comment (vivekkarnatiharness) [reply 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/email/EmailStepResponse.java:13]: Done

File: 860-orchestration-steps/src/main/java/io/harness/plancreator/steps/email/EmailStepNode.java (hunk 3 above)
Comment (vivekkarnatiharness) [reply 860-orchestration-steps/src/main/java/io/harness/plancreator/steps/email/EmailStepNode.java]: Done

File: 860-orchestration-steps/src/main/java/io/harness/plancreator/steps/email/EmailStepInfo.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+  @ApiModelProperty(dataType = SwaggerConstants.STRING_LIST_CLASSPATH)
+  @YamlSchemaTypes(value = {runtime})
+  ParameterField<List<TaskSelectorYaml>> delegateSelectors;
+
+  @Builder(builderMethodName = "infoBuilder")
+  public EmailStepInfo(ParameterField<String> to, ParameterField<String> cc, ParameterField<String> subject,
+      ParameterField<String> body, ParameterField<List<TaskSelectorYaml>> delegateSelectors) {
+    this.body = body;
+    this.to = to;
+    this.cc = cc;
```
[the remark this answers is not in the document]
Comment (vivekkarnatiharness) [reply 860-orchestration-steps/src/main/java/io/harness/plancreator/steps/email/EmailStepInfo.java:61]: done

**Review by puthrayaharness — APPROVED** [review]
> Reviewed 14 of 23 files at r2, 8 of 9 files at r3, 12 of 12 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34845)* status: 1. unresolved discussions.

**Review by puthrayaharness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34845)* status: :shipit: complete! ✓

**Review by puthrayaharness — APPROVED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34845)* status: :shipit: complete! ✓

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 13 of 23 files at r2, 7 of 9 files at r3, 10 of 12 files at r4, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34845)* status: 2 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`860-orchestration-steps/src/main/java/io/harness/serializer/OrchestrationStepsModuleRegistrars.java` line 177 at r5](https://reviewable.io/reviews/harness/harness-core/34845#-N672vDM8oQz8OqdlyWn:-N672vDM8oQz8OqdlyWo:bb500on) ([raw file](https://github.com/harness/harness-core/blob/80e1a66b482a70f459dc93fd9239a05c2cacf2ff/860-orchestration-steps/src/main/java/io/harness/serializer/OrchestrationStepsModuleRegistrars.java#L177)):*
> ```Java
>                    .yamlSchemaMetadata(YamlSchemaMetadata.builder()
>                                            .namespace(SchemaNamespaceConstants.PMS)
>                                            .modulesSupported(Arrays.asList(ModuleType.CD, ModuleType.PMS))
> ```

CI will not consume this?



_Code quote:_
```Java
modulesSupported(Arrays.asList(ModuleType.CD, ModuleType.PMS)
```
___
*[`860-orchestration-steps/src/main/java/io/harness/steps/email/EmailStepParameters.java` line 25 at r6](https://reviewable.io/reviews/harness/harness-core/34845#-N6736fn568DQn06NvR6:-N6736fn568DQn06NvR7:bv3vojp) ([raw file](https://github.com/harness/harness-core/blob/70bebd966b96314a089c5d977a5bab5efcc86943/860-orchestration-steps/src/main/java/io/harness/steps/email/EmailStepParameters.java#L25)):*
> ```Java
> @OwnedBy(HarnessTeam.CDC)
> @TypeAlias("EmailStepParameters")
> ```

Recaster alias is necessary

**Comment by vivekkarnatiharness** [comment]
*[`860-orchestration-steps/src/main/java/io/harness/serializer/OrchestrationStepsModuleRegistrars.java` line 177 at r5](https://reviewable.io/reviews/harness/harness-core/34845#-N672vDM8oQz8OqdlyWn:-N67I3Wl1dDplCDkcgDd:b-lsbjf9) ([raw file](https://github.com/harness/harness-core/blob/80e1a66b482a70f459dc93fd9239a05c2cacf2ff/860-orchestration-steps/src/main/java/io/harness/serializer/OrchestrationStepsModuleRegistrars.java#L177)):*
> CI will not consume this?
For now It is CD only. We will add it to CI later.

**Review by archit-harness — APPROVED** [review]
> Reviewed 5 of 5 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34845)* status: :shipit: complete! ✓

___
*[`860-orchestration-steps/src/main/java/io/harness/steps/email/EmailStepParameters.java` line 25 at r6](https://reviewable.io/reviews/harness/harness-core/34845#-N6736fn568DQn06NvR6:-N67mSSG2WbOb-bcXVTN:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/70bebd966b96314a089c5d977a5bab5efcc86943/860-orchestration-steps/src/main/java/io/harness/steps/email/EmailStepParameters.java#L25)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Recaster alias is necessary
</blockquote></details>

done

**Review by puthrayaharness — DISMISSED** [review]
> Reviewed 2 of 5 files at r7, 2 of 2 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34845)* status: :shipit: complete! ✓

**Comment by puthrayaharness** [comment]
trigger ti0

**Review by archit-harness — APPROVED** [review]
> Reviewed 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34845)* status: :shipit: complete! ✓

**Comment by vivekkarnatiharness** [comment]
trigger ti0

**Review by puthrayaharness — DISMISSED** [review]
> Reviewed 4 of 4 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34845)* status: :shipit: complete! ✓

**Review by prashantsharma04 — DISMISSED** [review]
> Reviewed 13 of 23 files at r2, 5 of 9 files at r3, 8 of 12 files at r4, 2 of 5 files at r7, 4 of 4 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34845)* status: :shipit: complete! ✓

**Review by archit-harness — APPROVED** [review]
> Reviewed 2 of 4 files at r10, 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34845)* status: :shipit: complete! ✓

**Review by puthrayaharness — APPROVED** [review]
> Reviewed 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34845)* status: :shipit: complete! ✓

**Pull request merged**